tel2puml/find_unique_graphs/otel_ingestion/json_data_converter.py

```diff
@@ -629,104 +631,148 @@ def process_header(
+    `list`[`dict`[`str`, `Any`]]
+    :return: The nested dictionary
+    :rtype: `dict`[`str`, `Any`]
+    """
+    result = parsed_data
     for key in reversed(path_segments[1:]):
         result = {key: result}
 
-    return result
+    if isinstance(result, dict):
+        return result
+    else:
+        raise TypeError(f"Expected type dict, got {type(result)} instead.")
 
 
 def _extract_simple_value(
-    data: dict[str, Any] | str, path: str
-) -> flatdict.FlatterDict | str:
+    data: dict[str, Any], path: str
+) -> dict[str, Any] | str | list[dict[str, Any]]:
     """Extract a simple value from JSON data using a path.
 
     :param data: The JSON data to traverse
     :type data: `dict`[`str`, `Any`]
     :param path: The simple path string
     :type path: `str`
     :return: The extracted value
-    :rtype: :class: `flatdict.FlatterDict` | `str`
+    :rtype: `dict`[`str`, `Any`] | `str` | `list`[`dict`[`str`, `Any`]]
     """
+    parsed_data: dict[str, Any] | str | list[dict[str, Any]] = data
     segments = path.split(":")
     for segment in segments:
-        data = _navigate_segment(data, segment)
+        if isinstance(parsed_data, dict):
+            parsed_data = _navigate_dict(parsed_data, segment)
 
-    if isinstance(data, dict) or isinstance(data, list):
-        return _flatten_json_dict(data)
-    return data
+    return parsed_data
 
 
-def _navigate_segment(
-    data: dict[str, Any] | str, segment: str
-) -> dict[str, Any] | str:
-    """Navigate through a segment of the JSON data.
+def _navigate_dict(
+    data: dict[str, Any], segment: str
+) -> dict[str, Any] | str | list[dict[str, Any]]:
+    """Navigate through a dictionary of the JSON data.
 
     :param data: The current data object
     :type data: `dict`[`str`, `Any`]
     :param segment: The segment to navigate
     :type segment: `str`
     :return: The value at the end of the navigation
-    :rtype: `dict`[`str`, `Any`] | `str`
+    :rtype: `dict`[`str`, `Any`] | `str` | `list`[`dict`[`str`, `Any`]]
     """
-    if isinstance(data, dict):
-        data = data[segment]
-    elif isinstance(data, list):
-        data = data[0]
+    try:
+        data_to_return = data[segment]
+        if isinstance(data_to_return, (dict, list, str)):
+            return data_to_return
+        else:
+            raise TypeError(
+                "Error navigating header - value should be of type dict,"
+                f" list or str - got {type(data[segment])} instead."
+            )
+    except KeyError:
+        raise KeyError(f"{segment} is not a valid key.")
 
-    return data
+
+def _navigate_list(
+    data: list[dict[str, Any]],
+) -> dict[str, Any]:
+    """Navigate through a list of the JSON data.
+
+    :param data: The current data object
+    :type data: `list`[`dict`[`str`, `Any`]]
+    :return: The value at index 0 of the list
+    :rtype: `dict`[`str`, `Any`]
+    """
+    try:
+        return data[0]
```

Comment:
why are we only taking the first entry from the list -will we not be skipping data?